src/libpgagroal/ev.c

```diff
@@ -0,0 +1,1667 @@
+   struct ev_periodic** w;
+   if (!target)
+   {
+      pgagroal_log_error("null pointer provided to stop\n");
+      return EV_ERROR;
+   }
+   ret = periodic_stop(ev, target);
+   list_delete(w, &ev->phead.next, target, ret);
+   return ret;
+}
+
+/*
+ * TODO: when developing config, make sure that the user has the backend that she claims to have.
+ * When it gets to this point, checks will not be made any longer, unless inside specific __ev_function.
+ * These functions will contain preprocessor checks so that the compilation does not break...
+ *
+ * int (*init)(struct ev_loop *);
+ * int (*loop)(struct ev_loop *);
+ * int (*io_start)(struct ev_loop*, struct ev_io*);
+ * int (*io_stop)(struct ev_loop*, struct ev_io*);
+ * int (*signal_init)(struct ev_loop*, struct ev_signal*);
+ * int (*signal_start)(struct ev_loop*, struct ev_signal*);
+ * int (*signal_stop)(struct ev_loop*, struct ev_signal*);
+ * int (*periodic_init)(struct ev_loop*, struct ev_periodic*);
+ * int (*periodic_start)(struct ev_loop*, struct ev_periodic*);
+ * int (*periodic_stop)(struct ev_loop*, struct ev_periodic*);
+ *
+ */
+static int
+setup_ops(struct ev_loop* ev)
+{
+   int ret = EV_OK;
+   bool mtt = ev->ctx.multithreading;
+
+   is_running = mtt ? __is_running_atomic : __is_running;
+   set_running = mtt ? __set_running_atomic: __set_running;
+   loop_break = mtt ? __break_atomic: __break;
+
+   if (ev->ctx.backend == EV_BACKEND_IO_URING)
+   {
+#if HAVE_URING
+      loop_init = __io_uring_init;
+      loop_fork = __io_uring_fork;
+      loop_destroy = __io_uring_destroy;
+      loop_start = __io_uring_loop;
+      io_start = __io_uring_io_start;
+      io_stop = __io_uring_io_stop;
+      periodic_init = __io_uring_periodic_init;
+      periodic_start = __io_uring_periodic_start;
+      periodic_stop = __io_uring_periodic_stop;
+      signal_start = __io_uring_signal_start;
+      signal_stop = __io_uring_signal_stop;
+#endif
+   }
+   else if (ev->ctx.backend == EV_BACKEND_EPOLL)
+   {
+#if HAVE_EPOLL
+      loop_init = __epoll_init;
+      loop_fork = __epoll_fork;
+      loop_destroy = __epoll_destroy;
+      loop_start = __epoll_loop;
+      io_start = __epoll_io_start;
+      io_stop = __epoll_io_stop;
+      periodic_init = __epoll_periodic_init;
+      periodic_start = __epoll_periodic_start;
+      periodic_stop = __epoll_periodic_stop;
+      signal_start = __epoll_signal_start;
+      signal_stop = __epoll_signal_stop;
+#endif
+   }
+   else if (ev->ctx.backend == EV_BACKEND_KQUEUE)
+   {
+      pgagroal_log_fatal("no support for kqueue yet!");
+   }
+
+   return ret;
+}
+
+static int
+setup_context(struct ev_context* ctx, struct configuration* config)
+{
+   /*
+    * TODO: should be an option.
+    */
+   ctx->multithreading = false;
+
+   /* set backend */
+
+   pgagroal_ev_print_backends();
+
+   if (!strlen(config->ev_backend))
+   {
+      pgagroal_log_warn("ev_backend not set in configuration file. Selected default: '%s'", FALLBACK_BACKEND);
```

Comment:
Please add a note about the configuration tunable `ev_backend`, so that the user knows how to configure it.
Also *add the configuration setting to the documentation* `CONFIGURATION.md` and friends.